### ISSUE TITLE
DWARF: Generate types with qualifier tags

### ISSUE THREAD

**Issue by ljmf00** [ldc-developers/ldc#3867]
Types with qualifiers like `const`, should generate `DW_TAG_const_type` instead of hardcoded `const(T)` basic type.

**Comment by ljmf00** [comment]
In order to add an `immutable` qualifier, LLVM upstream need to support `DW_TAG_immutable_type`.

**Comment by JohanEngelen** [comment]
Did you submit an upstream patch or issue? Can you link to it here?

**Comment by kinke** [comment]
`llvm::DIBuilder::createQualifiedType()` takes a `unsigned Tag`, so it might be possible to specify the DWARF value directly. There's also `DW_TAG_shared_type`; `inout` has no mapping AFAICT. No idea if/which of these qualifiers are supported for CodeView too.

**Comment by ljmf00** [comment]
> `llvm::DIBuilder::createQualifiedType()` takes a `unsigned Tag`, so it might be possible to specify the DWARF value directly. There's also `DW_TAG_shared_type`; `inout` has no mapping AFAICT.

Yes. I'm going to create a bug report and submit a patch anyway. This can be probably fixed independently here by defining that value manually, but functions like `isTypeTag` can't recognize it and bad behaviour may happen. I need to do it anyway in order for LLDB debuggers to understand it properly.

> No idea if/which of these qualifiers are supported for CodeView too.

I don't use CodeView, so I'm not aware of that. Also a note, `DW_TAG_immutable_type` was exclusively added for D in DWARF standard, AFAIK, so there is a small chance that this got implemented somewhere else, but don't take my words for granted.

**Comment by kinke** [comment]
> Also a note, DW_TAG_immutable_type was exclusively added for D in DWARF standard, AFAIK, so there is a small chance that this got implemented somewhere else, but don't take my words for granted.

Yep seen that it was added some 5 years ago for D. :) - The thing is that we emit DWARF-like LLVM debug infos, and LLVM converts that to DWARF or CodeView for us. Due to CodeView restrictions, there are already some DI special cases in place; these qualifiers might add some special cases on our side or require LLVM to handle custom values (or values not supported by CodeView) for us.

**Comment by ljmf00** [comment]
> > Also a note, DW_TAG_immutable_type was exclusively added for D in DWARF standard, AFAIK, so there is a small chance that this got implemented somewhere else, but don't take my words for granted.
> 
> Yep seen that it was added some 5 years ago for D. :) - The thing is that we emit generic DWARF-like LLVM debug infos, and LLVM converts that to DWARF or CodeView for us. Due to CodeView restrictions, there are already some DI special cases in place; these qualifiers might add some special cases on our side or require LLVM to handle custom values (or values not supported by CodeView) for us.

Is there any CodeView documentation you can reference to help me? If that is trivial, I can add support for it.

**Comment by kinke** [comment]
There's https://llvm.org/docs/PDB/CodeViewTypes.html, but AFAIK all reverse-engineered.

**Comment by ljmf00** [comment]
> There's [llvm.org/docs/PDB/CodeViewTypes.html](https://llvm.org/docs/PDB/CodeViewTypes.html), but AFAIK all reverse-engineered.

That seems the same situation for PDB. I wrote in https://github.com/microsoft/microsoft-pdb/issues/51 to request up-to-date information, but it seems a dead repository. Tools like `llvm-pdbutil` might help, as mentioned by an ex-Googler, who wrote PDB documentation and probably wrote that too. This can be hard to discover although, since I don't know any Microsoft language that uses such type qualifier.

**Comment by ljmf00** [comment]
> Did you submit an upstream patch or issue? Can you link to it here?

For reference: https://bugs.llvm.org/show_bug.cgi?id=52471

**Comment by ljmf00** [comment]
LLVM moved to Github. In the meanwhile, I closed https://github.com/llvm/llvm-project/issues/51813 . This tag should be available on the LLVM IR 14.0 .
